File: apps/admin/src/api/services/queue.ts
import apiClient from '../client'

export interface QueueEntry {
  id: string
  location_id: string
  customer_id: string
  appointment_id: string | null
  queue_number: string
  position: number
  estimated_wait_mins: number | null
  status: 'waiting' | 'called' | 'serving' | 'completed' | 'left'
  joined_at: string | null
  called_at: string | null
}

export interface QueueStatus {
  location_id: string
  total_waiting: number
  currently_serving: string | null
  entries: QueueEntry[]
}

export const queueApi = {
  getStatus: async (locationId: string): Promise<QueueStatus> => {
    const response = await apiClient.get(`/queue/status/${locationId}`)
    return response.data
  },

  checkIn: async (appointmentId: string): Promise<QueueEntry> => {
    const response = await apiClient.post('/queue/check-in', {
      appointment_id: appointmentId
    })
    return response.data
  },

  callNext: async (locationId: string): Promise<QueueEntry> => {
    const response = await apiClient.post(`/queue/call-next/${locationId}`)
    return response.data
  },

  callSpecific: async (entryId: string): Promise<QueueEntry> => {
    const response = await apiClient.post(`/queue/${entryId}/call`)
    return response.data
  },

  startServing: async (entryId: string): Promise<QueueEntry> => {
    const response = await apiClient.post(`/queue/${entryId}/start-serving`)
    return response.data
  },

  complete: async (entryId: string): Promise<QueueEntry> => {
    const response = await apiClient.post(`/queue/${entryId}/complete`)
    return response.data
  },

  markNoShow: async (entryId: string): Promise<QueueEntry> => {
    const response = await apiClient.post(`/queue/${entryId}/no-show`)
    return response.data
  },
}

File: apps/api/app/schemas/queue.py
from pydantic import BaseModel
from uuid import UUID
from datetime import datetime
from typing import Optional

class CheckInRequest(BaseModel):
    appointment_id: UUID

class WalkInRequest(BaseModel):
    location_id: UUID

class QueueEntryResponse(BaseModel):
    id: UUID
    location_id: UUID
    customer_id: UUID
    appointment_id: Optional[UUID]
    queue_number: str  # Formatted as "A001"
    position: int
    estimated_wait_mins: Optional[int]
    status: str
    joined_at: Optional[datetime]
    called_at: Optional[datetime]

    class Config:
        from_attributes = True

class QueuePositionResponse(BaseModel):
    queue_number: str
    position: int
    estimated_wait_mins: int
    status: str
    total_ahead: int

class QueueEntryDict(BaseModel):
    id: str
    queue_number: str
    position: int
    status: str
    estimated_wait_mins: Optional[int]
    joined_at: Optional[str]

class QueueStatusResponse(BaseModel):
    location_id: str
    total_waiting: int
    currently_serving: Optional[str]
    entries: list[dict]
